### DISCUSSION TITLE
Dispose services

### DISCUSSION THREAD

**Discussion by Nuthin2do** [CodinGame/monaco-vscode-api#523] [Q&A] [answered]
Hi again,

i searched the repo and the demo but didn't find anything which could be used to unregister or stop the services running i initialize at the beginning of my application.

```
private async initWorkers() {
    type WorkerLoader = () => Worker;

    const workerLoader: Partial<Record<string, WorkerLoader>> = {
      TextEditorWorker: () =>
        new Worker(
          new URL(
            "monaco-editor/esm/vs/editor/editor.worker.js",
            import.meta.url
          ),
          { type: "module" }
        ),
      TextMateWorker: () =>
        new Worker(
          new URL(
            "@codingame/monaco-vscode-textmate-service-override/worker",
            import.meta.url
          ),
          { type: "module" }
        ),
      TypescriptWorker: () =>
        new Worker(
          new URL(
            "@codingame/monaco-vscode-standalone-typescript-language-features/worker",
            import.meta.url
          ),
          { type: "module" }
        ),
    };

    window.MonacoEnvironment = {
      getWorker: function (moduleId, label) {
        const workerFactory = workerLoader[label];
        if (workerFactory != null) {
          return workerFactory();
        }
        throw new Error(`Unimplemented worker ${label} (${moduleId})`);
      },
    };
  }

  public async initializeSetup(
    id: string,
    editorOptions: editor.IStandaloneDiffEditorConstructionOptions
  ) {
    await this.initWorkers();

    await initUserConfiguration(JSON.stringify(defaultConfiguration));

    await initialize({
      ...getModelServiceOverride(),
      ...getThemeServiceOverride(),
      ...getLanguagesServiceOverride(),
      ...getTextMateServiceOverride(),
      ...getConfigurationServiceOverride(),
      ...getDebugServiceOverride(),
    });

    const container = document.getElementById(id);
    if (container) {
      this.editor = editor.create(container, { model: null });
      this.editor.layout();
      this.registerBreakpointEvent(container);
      this.registerEditorEvents();
    }
  }
```

In my application if i navigate to the script section once then everything runs fine and i can start coding in the editor. When i navigate to another section of my application i get disposed of all modelRef, the editor instance and the fileSystemOverlay. As soon as i navigate to the script section again the application crashes and i get an error like '**Services are already initialized**'. That means getting rid of the editor and the container the editor is attached being destroyed is not enough. How can i ensure everything is cleaned up when after using the editor (i mean when i navigate away from the editor).

Thank you very much for your answer.

**Comment by CGNonofr** [comment, marked as answer]
Unfortunately, that's not something that is possible for multiples reasons:
- This is not needed for VSCode, so no effort was done toward that direction
- Some services register global stuff (like onbeforeunload, or create workers....) and nothing exists to unregister them
- Most of the code is actually what is called "contributions". Some classes in which services are injected and that can't be unloaded

So I see 2 solutions:
- if you don't need the service configuration to change: you can just make sure your services are only initialized once before the first time you need it
- if you want to update the service configuration: you have no other choice than reloading the page. It includes the possibility to load VSCode inside an iframe that is loaded each time.

**Reply by Nuthin2do** [reply]
I think loading VSCode inside an iframe is the best solution. Thank you for your answer.

**Reply by vitor-txt** [reply]
How can I clear the memory after closing the iframe? I'm facing a problem with closing and reopening the iframe with a VSCode and I don't know how to proceed. My first approach was to try to implement the disposal, but seems that is not working properly in my tests:
![Memory leak test](https://github.com/user-attachments/assets/4bd73579-f16d-4e6e-be4f-6a441dff514e)
Snapshots taken:
1. Before the first open
2. With VSCode opened in the iframe
3. After dispose
4. Two minutes after closing the iframe
5. Opening the iframe a second time
6. After the second dispose
7. Two minutes after closing the iframe
8. The third time opened

**Reply by CGNonofr** [reply]
I'm not sure what you mean. If you remove the iframe, everything is cleaned
